packages/protocol/docs/how_taiko_proves_blocks.md

```diff
@@ -0,0 +1,361 @@
+
+- `blockhash`: EVM allows access to the most recent 256 block hashes. All these hashes are available inside the plonk lookup table. ZKP must prove that the lookup table is consistent with L2's historical values.
+  - `blockhash(block.number - 1)`, has the same value as in the block header and is also the same value as the parent block's hash on L2.
+  - The other 255 hashes, `blockhash(block.number - 256)` to `blockhash(block.number - 2)` are checked in the anchor transaction to simplify circuits. Therefore, as long as the anchor transaction is zk-proven, these 255 ancestor hashes are proven indirectly.
+- `block.basefee`: verified to be the correct value in the anchor transaction.
+- `block.chainid`: this field is also checked by the anchor transaction, so no extra ZKP circuits are required.
+- `block.coinbase`: ZKP must verify the value must be the same as `meta.beneficiary`. Again, the metadata hash is part of the ZK instance.
+- `block.difficulty`: this is now the same as `block.prevrandao`, so we only check `block.prevrandao`.
+- `block.gaslimit`: ZKP must verify this value must equal `meta.gasLimit`.
+- `block.number`: this must be checked against the block header and `meta.id`.
+- `block.prevrandao`: this must be checked against the `mixHash` field in the L2 block header and `meta.mixHash`.
+- `block.timestamp`: this must be checked against the `timestamp` field in the L2 block header and `meta.proposedAt`.
+
+### Block Header
+
+Not all block header data is available in the L1 contracts; therefore, the ZKP must verify the integrity of the block header and ensure the block header hashes into the same value as `evidence.blockHash`, which is part of the ZK instance input.
+
+```solidity
+struct BlockHeader {
+  bytes32 parentHash;
+  bytes32 ommersHash;
+  address beneficiary;
+  bytes32 stateRoot;
+  bytes32 transactionsRoot;
+  bytes32 receiptsRoot;
+  bytes32[8] logsBloom;
+  uint256 difficulty;
+  uint128 height;
+  uint64 gasLimit;
+  uint64 gasUsed;
+  uint64 timestamp;
+  bytes extraData;
+  bytes32 mixHash;
+  uint64 nonce;
+  uint256 baseFeePerGas;
+  bytes32 withdrawalsRoot;
+}
+```
+
+In addition, ZKP must also prove the following:
+
+- `parentHash` must be the same as `evidence.parentHash`.
+- `ommersHash` must be the keccak256 of `[]`,or `0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347`.
+- `beneficiary` must be `meta.beneficiary` (duplicated, as stated above).
+- `logsBloom` must be a `bytes32[8]` with all zeros.
+- `difficulty` == 0.
+- `height` must be `meta.id`.
+- `gasLimit` == `meta.gasLimit`.
+- `gasUsed` must be the sum of all gas used by all valid transactions in the block.
+- `timestamp` == `meta.proposedAt`.
+- `extraData` == "".
+- `mixHash` == `meta.mixHash`.
+- `nonce` == 0.
+- `baseFeePerGas` == the calculated EIP-1559 style base fee.
```

Comment:
And it needs to also prove that all base fee was transferred to `treasury` address?

Reply:
Added a small section for it.